certora/funnels/automation/StableDepositorUniV3.spec

```diff
@@ -0,0 +1,509 @@
+
+    assert wardsAfter == wardsBefore, "setConfig did not keep unchanged every wards[x]";
+    assert budsAfter == budsBefore, "setConfig did not keep unchanged every buds[x]";
+    assert numGem0Gem1After == to_mathint(num), "setConfig did not set configs[gem0][gem1][fee][tickLower][tickUpper].num to num";
+    assert zzzGem0Gem1After == 0, "setConfig did not set configs[gem0][gem1][fee][tickLower][tickUpper].zzz to 0";
+    assert amt0Gem0Gem1After == to_mathint(amt0), "setConfig did not set configs[gem0][gem1][fee][tickLower][tickUpper].amt0 to amt0";
+    assert amt1Gem0Gem1After == to_mathint(amt1), "setConfig did not set configs[gem0][gem1][fee][tickLower][tickUpper].amt1 to amt1";
+    assert req0Gem0Gem1After == to_mathint(req0), "setConfig did not set configs[gem0][gem1][fee][tickLower][tickUpper].req0 to req0";
+    assert req1Gem0Gem1After == to_mathint(req1), "setConfig did not set configs[gem0][gem1][fee][tickLower][tickUpper].req1 to req1";
+    assert hopGem0Gem1After == to_mathint(hop), "setConfig did not set configs[gem0][gem1][fee][tickLower][tickUpper].hop to hop";
+    assert numOtherAfter == numOtherBefore, "setConfig did not keep the rest of configs[x][y][z][a][b].num";
+    assert zzzOtherAfter == zzzOtherBefore, "setConfig did not keep the rest of configs[x][y][z][a][b].zzz";
+    assert amt0OtherAfter == amt0OtherBefore, "setConfig did not keep the rest of configs[x][y][z][a][b].amt0";
+    assert amt1OtherAfter == amt1OtherBefore, "setConfig did not keep the rest of configs[x][y][z][a][b].amt1";
+    assert req0OtherAfter == req0OtherBefore, "setConfig did not keep the rest of configs[x][y][z][a][b].req0";
+    assert req1OtherAfter == req1OtherBefore, "setConfig did not keep the rest of configs[x][y][z][a][b].req1";
+    assert hopOtherAfter == hopOtherBefore, "setConfig did not keep the rest of configs[x][y][z][a][b].hop";
+}
+
+// Verify revert rules on setConfig
+rule setConfig_revert(address gem0, address gem1, uint24 fee, int24 tickLower, int24 tickUpper, int32 num, uint32 hop, uint96 amt0, uint96 amt1, uint96 req0, uint96 req1) {
+    env e;
+
+    mathint wardsSender = wards(e.msg.sender);
+
+    setConfig@withrevert(e, gem0, gem1, fee, tickLower, tickUpper, num, hop, amt0, amt1, req0, req1);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = wardsSender != 1;
+    bool revert3 = gem0 >= gem1;
+
+    assert revert1 => lastReverted, "revert1 failed";
+    assert revert2 => lastReverted, "revert2 failed";
+    assert revert3 => lastReverted, "revert3 failed";
+    assert lastReverted => revert1 || revert2 || revert3, "Revert rules are not covering all the cases";
+}
+
+// Verify correct storage changes for non reverting deposit
+rule deposit(address gem0, address gem1, uint24 fee, int24 tickLower, int24 tickUpper, uint128 amt0Min, uint128 amt1Min) {
```

Comment:
The proofs are taking forever to run so haven't confirmed yet, but I _think_ `deposit`, `withdraw`, and `collect here are missing any verification that the corresponding external call takes place. Again, ofc unit tests should be covering that, but you could e.g. use a helper contract to record and then confirm the values passed if you wanted to be more complete.

Reply:
Let me see if I can do this with ghost variables, as I'm using summarization to avoid actual mock contracts.

Reply:
Just pushed a change proving with ghost variables, seems to be working fine, it has thrown a violation if I comment the external code in the automation contract. Evaluating if extending this idea to save each field and compare, but for now at least it proves the external call exists.

Reply:
I pushed a new commit checking field by field. The only one that is not perfectly done is `data` for the stableSwapper, as it doesn't seem to allow me to save the `bytes` param. For now checking the length. I guess it is more than enough.

Reply:
Nice, I did not know that summaries were so powerful now.